### PR TITLE
Add annotation to mark a type as DoNotMock

### PR DESCRIPTION
**Pull request by TimvdLippe**
Mocking types that users not own [1] or are severely complicating test
logic [2] leads to brittle or wrong tests. In particular, the
StackOverflow answer is wrong, as the contract of java.util.Map is
violated. When a new key is added to the Map, the stubbed return would be wrong.
In Google we have used the DoNotMock annotation via ErrorProne [3]
to annotate these types, as well as an internal list of types that can't
be mocked (this includes several java.util types). We are using a custom
Mockmaker to enforce this on run-time.

Based on our successful experience with DoNotMock (we have seen a large
reduction in bad/broken tests for types involved), we are proposing to
open source this into Mockito itself.

The DoNotMock annotation can be added to any type, e.g. classes and
interfaces. If, in the type hierarchy of the class-to-be-mocked, there
is a type that is annotated with DoNotMock, Mockito will throw a
DoNotMockException.

This would help preventing issues such as #1827 and #1734 which is
in-line with the guidance on our wiki [1]. A follow-up change would
allow us to define external types (like the java.util types) that can't
be mocked. (We can't add the annotation to the types, as they live in the
JDK instead.)

[1]: https://github.com/mockito/mockito/wiki/How-to-write-good-tests#dont-mock-a-type-you-dont-own
[2]: https://stackoverflow.com/a/15820143
[3]: https://errorprone.info/api/latest/com/google/errorprone/annotations/DoNotMock.html

### CODE REVIEW

**Comment by mockitoguy** [comment]
Great submission, thank you!

> we are proposing to open source this into Mockito itself.

Fantastic idea. THANK you!

> We are using a custom Mockmaker to enforce this on run-time.

Wow! Interesting use of this API. Can you enumerate other interesting use cases Google has extended Mockito behavior via our public API? I'd love us to offer first class support for such features.

> A follow-up change would
allow us to define external types (like the java.util types) that can't
be mocked.

+1

**Comment by mockitoguy** [comment]
Request: can we add a new method to override this just in case? Here's a use case: let's say someone incorrectly sets DoNotMock annotation on a type and ships the code in a library. The consumers of the library don't have a good way to remove DoNotMock annotation if they don't own the library code. Thoughts?

Example API:

```java
mock(Foo.class, withSettings().overrideDoNotMock())
```

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (mockitoguy) [inline src/main/java/org/mockito/internal/MockitoCore.java]: Do we want to honor this specific annotation (Mockito API) or _any_ annotation with name "DoNotMock".

The latter seems more useful. For example: a library owner can mark types with any "DoNotMock" annotation without adding compile dependency on mockito-core.

File: src/test/java/org/mockitousage/annotation/DoNotMockTest.java
Comment (mockitoguy) [inline src/test/java/org/mockitousage/annotation/DoNotMockTest.java]: Great test coverage.

**Comment by TimvdLippe** [comment]
> Wow! Interesting use of this API. Can you enumerate other interesting use cases Google has extended Mockito behavior via our public API? I'd love us to offer first class support for such features.

Yes I am auditing/discovering usages of Mockito API and reviewing them whether they are worth keeping (and open sourcing) or should be changed. This is one of many :smile: 

> Request: can we add a new method to override this just in case? Here's a use case: let's say someone incorrectly sets DoNotMock annotation on a type and ships the code in a library. The consumers of the library don't have a good way to remove DoNotMock annotation if they don't own the library code. Thoughts?

Yes we already have the concept of "grandfathered" mock violations. E.g. all existing mocks of a type that you want to add `@DoNotMock` to will not immediately break. This allows us to put a hold on new violations while users can (over time) remove violations. I am still trying to come up with a scalable solution for implementing that logic, which will solve your use case as well.

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (TimvdLippe) [reply src/main/java/org/mockito/internal/MockitoCore.java]: I would like to keep this for now and open it up later to match on name. If we would implement that now, I would break all users of ErrorProne that rely on `@DoNotMock` but have grandfathered mock violations.

If we can gather feedback first, we can see adoption patterns and change it to match on the name later. WDYT?

Comment (mockitoguy) [reply src/main/java/org/mockito/internal/MockitoCore.java]: How about a happy medium so that we don't need to revisit this in the future (and cause breaking changes :) - we can match for "org.mockito.DoNotMock" by String. This should be backwards compatible for G. and other companies / projects. Thoughts?

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (mockitoguy) [inline src/main/java/org/mockito/internal/MockitoCore.java]: I played with the code and noticed that "mockTypeToCheck != null" part of this statement is either: a) not needed or b) not covered with tests.

Do we need this part?

**Comment by raphw** [comment]
Really interesting feature, also with regards to returning mocks from methods, I think. But very much +1 on the override functionality. There is always a "but" that might need consideration.

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (TimvdLippe) [reply src/main/java/org/mockito/internal/MockitoCore.java]: It is covered by mocking an interface that directly has the `DoNotMock` annotation. I added a test to cover this case.

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (TimvdLippe) [reply src/main/java/org/mockito/internal/MockitoCore.java]: Fixed

**Comment by TimvdLippe** [comment]
It is now possible to provide a custom `DoNotMockEnforcer` where you have full control over what types can be mocked. See the `extTest` test for an example.

Given the positive response, I will now spend some time integrating the changes in Google and see if it all builds. Given that is all positive and you have no more review comments, I will merge this PR.

**Comment by mockitoguy** [comment]
> It is now possible to provide a custom DoNotMockEnforcer

Curious, are you going to use this new plugin at G? If we don't have a user, I suggest we avoid complexity (it can be added later if needed).

Thank you for implementing name matching for the annotation!!!

**Comment by TimvdLippe** [comment]
> > It is now possible to provide a custom DoNotMockEnforcer
> 
> Curious, are you going to use this new plugin at G? If we don't have a user, I suggest we avoid complexity (it can be added later if needed).

Yes we will immediately use it. We have some additional classes that we are annotating as well as our "grandfathering violations" system.

Note that there is no other nice way of overriding the method as requested in https://github.com/mockito/mockito/pull/1833#issuecomment-559627262 (a method call per mock creation would mean that we would need to change every existing mock in our codebase to include it, which is very cumbersome)
> Thank you for implementing name matching for the annotation!!!

Glad you like it :smile:

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (mockitoguy) [inline src/main/java/org/mockito/internal/MockitoCore.java]: I suggest to drop this conditional, type cannot be null (unless I'm wrong :-)

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (mockitoguy) [inline src/main/java/org/mockito/internal/MockitoCore.java]: I suggest "MOCKABLE_CLASSES"

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (mockitoguy) [inline src/main/java/org/mockito/internal/MockitoCore.java]: I feel uneasy about this perf tweak. Do we need it? Does it really make things faster? Static state that keeps growing with every mocked type may bite us later.

Did you find this perf tweak useful at G?

Comment (TimvdLippe) [reply src/main/java/org/mockito/internal/MockitoCore.java]: I am pretty sure this is optimization is necessary as we can have large type hierarchies. I have asked internally for more background, but I expect this optimization to be necessary to not have a large performance impact for existing applications.

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (TimvdLippe) [reply src/main/java/org/mockito/internal/MockitoCore.java]: This will be `null` for `.getSuperClass()` of interfaces and `Object.class`. The tests will fail if we remove it.

File: src/main/java/org/mockito/plugins/DoNotMockEnforcer.java
Comment (TimvdLippe) [reply src/main/java/org/mockito/plugins/DoNotMockEnforcer.java]: I felt like this was the nicer way of handling state, since we have to stop type traversal. I will mark it as Incubating and might change later, depending on how well it integrates with our internal system.

File: src/main/java/org/mockito/internal/configuration/DefaultDoNotMockEnforcer.java
Comment (TimvdLippe) [reply src/main/java/org/mockito/internal/configuration/DefaultDoNotMockEnforcer.java]: Yes that's better :)

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (TimvdLippe) [reply src/main/java/org/mockito/internal/MockitoCore.java]: :+1:

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (mockitoguy) [reply src/main/java/org/mockito/internal/MockitoCore.java]: Adding perf tweaks "on hunch", without evidence is not good for the codebase (growing complexity :-). I don't mind this change but I'd love to get your assurance that this is useful.

You have the approval so ship this change at will!

Comment (TimvdLippe) [reply src/main/java/org/mockito/internal/MockitoCore.java]: I will double check before merging. I haven't had time during the Christmas vacation to do this work. I probably get to this late next week.

**Comment by TimvdLippe** [comment]
I have done an initial pass at integrating this internally with good results! We are doing additional verification to confirm everything is working as expected, but thus far it is working out.

**Comment by grimreaper** [comment]
Hey, this is a fantastic feature and I'd love for this to make it in. Can I gently nudge the and reviewers to finish this up?

**Comment by TimvdLippe** [comment]
I still have to integrate this in the Google monorepository and figure out whether it is working correctly. Sadly I haven't had the time to do so.

**Comment by TimvdLippe** [comment]
#2026 is another instance where we could benefit from `DoNotMock` to prevent users from running into problematic situations.

**Comment by TimvdLippe** [comment]
There are a fair few issues in our backlog about mocking classes like `File`, `Method` and `Map`, all of which cause undefined behavior in Mockito internals. Therefore, I think it is worth reviving this PR and see if we can land this to avoid people running into this problem. It would be a breaking change though, so we probably need an incubation timeline to get people to test it out.

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (bric3) [reply src/main/java/org/mockito/internal/MockitoCore.java:86]: Maybe leave this as a comment

File: src/main/java/org/mockito/DoNotMock.java
Comment (bric3) [inline src/main/java/org/mockito/DoNotMock.java:25]: It would be nice to add a simple code snippet with the use of the reason. 

Since annotation are likely to be how user discover this feature, it may makes sense to me to also describe how to make non-mockable the non-annotatable types like `File`.

File: src/main/java/org/mockito/internal/configuration/DefaultDoNotMockEnforcer.java
Comment (bric3) [inline src/main/java/org/mockito/internal/configuration/DefaultDoNotMockEnforcer.java:19]: Shouldn't it be `@org.mockito.DoNotMock` ? Maybe adding the actual annotation might be useful as well.

File: src/main/java/org/mockito/DoNotMock.java
Comment (bric3) [inline src/main/java/org/mockito/DoNotMock.java:35]: Maybe rename this as "reason", does an "alternative" field could make sense ?

**Review by bric3 — APPROVED** [review]
This is a nice feature in my opinion. And it can really help improve the quality of large code base.

About third party types, could it make sense to have a programmatic way to fill the non-mockable types in the default plugin implementation.

File: src/main/java/org/mockito/DoNotMock.java
Comment (TimvdLippe) [reply src/main/java/org/mockito/DoNotMock.java:35]: If we used `value`, then we can have `@DoNotMock("Special reason")` rather than `@DoNotMock(reason = "Special reason")`. Still, I think `reason` is clearer and also is better to read when reading the whole annotation usage. 👍

File: src/main/java/org/mockito/internal/configuration/DefaultDoNotMockEnforcer.java
Comment (TimvdLippe) [reply src/main/java/org/mockito/internal/configuration/DefaultDoNotMockEnforcer.java:19]: I think this is the official notation, but I have changed it to yours as that is probably more intuitive.

File: src/main/java/org/mockito/DoNotMock.java
Comment (TimvdLippe) [reply src/main/java/org/mockito/DoNotMock.java:25]: I have added an example, good suggestion 👍 

I think we can ship this without adding classes like `File` in Mockito. Then once we know the enforcer is working as expected, we can make the switch in the next major version of Mockito to by default disallow mocking these classes.

File: src/main/java/org/mockito/internal/MockitoCore.java
Comment (TimvdLippe) [reply src/main/java/org/mockito/internal/MockitoCore.java:86]: Done

**Comment by TimvdLippe** [comment]
Thanks everyone for your input and feedback! I will be publishing Mockito 4.1.0 that includes this feature 🎉

**Pull request merged**
